xds/client.go

```diff
@@ -0,0 +1,199 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/gojek/courier-go/xds/backoff"
+	"github.com/gojek/courier-go/xds/log"
+)
+
+// stream is used to mock reloadingStream
+type stream interface {
+	Send(req *v3discoverypb.DiscoveryRequest) error
+	Recv() (*v3discoverypb.DiscoveryResponse, error)
+	startReloader(ctx context.Context)
+}
+
+// Options specifies options to be provided for initialising the xds client
+type Options struct {
+	XDSTarget       string
+	NodeProto       *v3corepb.Node
+	ClientConn      grpc.ClientConnInterface
+	BackoffStrategy backoff.Strategy
+	Logger          log.Logger
+}
+
+// NewClient returns a new ADS client stream using the *grpc.ClientConn provided.
+func NewClient(opts Options) (*Client, error) {
+	opts = setDefaultOpts(opts)
+
+	c := &Client{
+		xdsTarget:   opts.XDSTarget,
+		cc:          opts.ClientConn,
+		nodeProto:   opts.NodeProto,
+		done:        make(chan struct{}),
+		receiveChan: make(chan []*v3endpointpb.ClusterLoadAssignment),
+		log:         opts.Logger,
+	}
+
+	cc := opts.ClientConn
+	rs := &reloadingStream{
+		reloadCh:    make(chan error, 1),
+		onReConnect: c.onStreamReConnect,
+		connTimeout: 10 * time.Second,
+		log:         opts.Logger,
+		strategy:    opts.BackoffStrategy,
+		ns: func() v3discoverypb.AggregatedDiscoveryServiceClient {
+			return v3discoverypb.NewAggregatedDiscoveryServiceClient(cc)
+		},
+	}
+
+	c.stream = rs
+
+	return c, rs.createStream()
+}
+
+// Client performs the actual ADS RPCs using the ADS v3 API. It creates an
+// ADS stream on which the xdsTarget resources are received.
+type Client struct {
+	nodeProto *v3corepb.Node
+	cc        grpc.ClientConnInterface
+	stream    stream
+	log       log.Logger
+
+	xdsTarget  string
+	vsn, nonce string
+
+	done        chan struct{}
+	receiveChan chan []*v3endpointpb.ClusterLoadAssignment
+}
+
+// Receive returns a channel where ClusterLoadAssignment resource updates can be received
+func (c *Client) Receive() <-chan []*v3endpointpb.ClusterLoadAssignment {
+	return c.receiveChan
+}
+
+// Done returns a channel which is closed when the run loop stops due to context expiry
+func (c *Client) Done() <-chan struct{} {
+	return c.done
+}
+
+// Start will wait updates from control plane, it is non-blocking
+func (c *Client) Start(ctx context.Context) error {
+	c.log.Info("xds: Starting ads client", "node", c.nodeProto, "target", c.xdsTarget)
+
+	go c.stream.startReloader(ctx)
+	go c.run(ctx)
+
+	return c.sendRequest(nil)
+}
+
+func (c *Client) run(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			close(c.done)
+			close(c.receiveChan)
+
+			return
+		default:
+			resp, err := c.stream.Recv()
+			if err != nil {
+				c.log.Error(err, "xds: error stream.Recv()")
+
+				break
+			}
+
+			if url := resp.GetTypeUrl(); url != resource.EndpointType {
+				c.nack(fmt.Errorf("xds: resource type (%s) is not EndpointResource in server response", url))
+
+				break
```

Comment:
If we break here, run function terminates. Do we want this behaviour here and why?

Reply:
Same as https://github.com/gojek/courier-go/pull/3#discussion_r851591859